plone/formwidget/namedfile/widget.rst

```diff
@@ -555,6 +555,249 @@ being returned::
+
+  >>> file_widget.request = TestRequest(form={'widget.name.file': upload, 'widget.name.file.action': 'replace'})
+  >>> file_widget.update()
+  >>> uploaded = file_widget.extract()
+  >>> uploaded
+  <ZPublisher.HTTPRequest.FileUpload instance at ...>
+
+  >>> content.file_field = ascii_file_converter.toFieldValue(uploaded)
+  >>> content.file_field
+  'filenameb64:cGxvbmUucGRm;datab64:cmFuZG9tIGZpbGUgY29udGVudA=='
+
+
+  >>> data = cStringIO.StringIO('no image')
+  >>> field_storage = FieldStorageStub(data, filename='logo.tiff')
+  >>> upload = FileUpload(field_storage)
+
+  >>> image_widget.request = TestRequest(form={'widget.name.image': upload, 'widget.name.image.action': 'replace'})
+  >>> image_widget.update()
+  >>> uploaded = image_widget.extract()
+  >>> uploaded
+  <ZPublisher.HTTPRequest.FileUpload instance at ...>
+
+  >>> content.image_field = ascii_file_converter.toFieldValue(uploaded)
+  >>> content.image_field
+  'filenameb64:bG9nby50aWZm;datab64:bm8gaW1hZ2U='
+
+
+Prepare for a new request cycle::
+
+  >>> file_widget = setup_widget('file', content, True)
+  >>> image_widget = setup_widget('image', content, True)
+
+
+The new image/file shows up in the rendered widget::
+
+  >>> file_widget.update()
+  >>> print(file_widget.render())
+  <... class="named-file-widget required ascii-field" id="widget.id.file">...
+  <a href="http://127.0.0.1/++widget++widget.name.file/@@download/plone.pdf">plone.pdf</a>...
+  <input type="radio"... id="widget.id.file-nochange"...
+  <input type="radio"... id="widget.id.file-replace"...
+  <input type="file"... id="widget.id.file-input"...
+
+  >>> image_widget.update()
+  >>> print(image_widget.render())
+  <... class="named-image-widget required ascii-field" id="widget.id.image">...
+  <img src="http://127.0.0.1/++widget++widget.name.image/@@download/logo.tiff"...
+  <a href="http://127.0.0.1/++widget++widget.name.image/@@download/logo.tiff">logo.tiff</a>...
+  <input type="radio"... id="widget.id.image-nochange"...
+  <input type="radio"... id="widget.id.image-replace"...
+  <input type="file"... id="widget.id.image-input"...
+
+
+Prepare for a new request cycle::
+
+  >>> file_widget = setup_widget('file', content)
+  >>> image_widget = setup_widget('image', content)
+
+#  >>> interact(locals())
+
+Resubmit, but keep the data::
+
+  >>> file_widget.request = TestRequest(form={'widget.name.file': '', 'widget.name.file.action': 'nochange'})
+  >>> file_widget.update()
+  >>> uploaded = file_widget.extract()
+  >>> uploaded
+  <plone.namedfile.file.NamedFile object at ...>
+
+  >>> content.file_field = ascii_file_converter.toFieldValue(uploaded)
+  >>> content.file_field
+  'filenameb64:cGxvbmUucGRm;datab64:cmFuZG9tIGZpbGUgY29udGVudA=='
+
+
+  >>> image_widget.request = TestRequest(form={'widget.name.image': '', 'widget.name.image.action': 'nochange'})
+  >>> image_widget.update()
+  >>> uploaded = image_widget.extract()
+  >>> uploaded
+  <plone.namedfile.file.NamedFile object at ...>
```

Comment:
Should rather be ``<plone.namedfile.file.NamedImage object at ...>``, and is if I do a ``image_widget.value = None; image_widget.update()`` instead of ``setup_widget``...
I don't care about that after hours of z3c.form horror.